Creating dedicated memory pools for data and control traffic

rado17 opened the pull request:
Create dedicated memory pools for Wi-Fi management and
data operations

rado17 committed:
drivers: wifi: Create dedicated mem pool for Wi-Fi driver

Create dedicated memory pools for Wi-Fi management and
data operations (defaults: 20KB for management and 130KB for data).
Setting Data pool to 110KB for non-Nordic SOCs to resolve
RAM overflows seen in twister runs.

Remove the `HEAP_MEM_POOL_ADD_SIZE_NRF70` hint since we are
creating separate heaps for driver and not allocating from
system heap.

Signed-off-by: Ravi Dondaputi <ravi.dondaputi@nordicsemi.no>

krish2718 commented on drivers/wifi/nrfwifi/Kconfig.nrfwifi:
```diff
@@ -541,6 +541,15 @@ config NRF70_RSSI_STALE_TIMEOUT_MS
 if NETWORKING
 # Finetune defaults for certain system components used by the driver
+
+config NRF_WIFI_CTRL_HEAP_SIZE
```
in order to leverage the newly added per-module HEAP hints, may be we should use `HEAP_MEM_POOL_ADD_SIZE_NRF70_CTRL` and ``HEAP_MEM_POOL_ADD_SIZE_NRF70_DATA`, this way there won't be any changes to the samples/applications. WDYT?

rado17 replied:
We are creating a separate pool for each of the control path and data path, and not allocating from system heap. Since `HEAP_MEM_POOL_ADD_SIZE_*` is used just as a hint to verify the available system heap, using this name convention here may not be appropriate.

krish2718 replied:
ok, then existing `HEAP_MEM_POOL_ADD_SIZE_` have to be removed or reduced, no? And then for `NRF_WIFI_DATA_HEAP_SIZE` needs to be adjusted per-sample as per the network buffer configuration?

rado17 committed:
manifest: nrf_wifi: Pull in changes for dedicated memory pools

Pull changes for creating separate memory pools for control traffic
and data traffic.

Signed-off-by: Ravi Dondaputi <ravi.dondaputi@nordicsemi.no>

krish2718 commented on drivers/wifi/nrf_wifi/Kconfig.nrfwifi:
```diff
@@ -552,6 +546,15 @@ config NRF70_RSSI_STALE_TIMEOUT_MS
 if NETWORKING
 # Finetune defaults for certain system components used by the driver
+
+config NRF_WIFI_CTRL_HEAP_SIZE
+	int "Dedicated memory pool for control path"
+	default 21000
+
+config NRF_WIFI_DATA_HEAP_SIZE
+	int "Dedicated memory pool for data plane"
+	default 129000
```
this would break scan-only, need a lower heap for that.

rado17 replied:
Allocated 6kB for SCAN_ONLY case.

krish2718 commented on modules/nrf_wifi/shim.c:
```diff
@@ -27,21 +28,79 @@
 #include "qspi_if.h"
 
 LOG_MODULE_REGISTER(wifi_nrf, CONFIG_WIFI_NRF70_LOG_LEVEL);
+#if defined(CONFIG_NOCACHE_MEMORY)
+K_HEAP_DEFINE_NOCACHE(wifi_drv_ctrl_mem_pool, CONFIG_NRF_WIFI_CTRL_HEAP_SIZE);
+K_HEAP_DEFINE_NOCACHE(wifi_drv_data_mem_pool, CONFIG_NRF_WIFI_DATA_HEAP_SIZE);
+#else
+K_HEAP_DEFINE(wifi_drv_ctrl_mem_pool, CONFIG_NRF_WIFI_CTRL_HEAP_SIZE);
+K_HEAP_DEFINE(wifi_drv_data_mem_pool, CONFIG_NRF_WIFI_DATA_HEAP_SIZE);
+#endif /* CONFIG_NOCACHE_MEMORY */
```
System heap uses cached heap `K_HEAP_DEFINE`, any reason to use uncached? In some platforms it can impact performance, like 54H.

rado17 replied:
Since this is protected with `CONFIG_NOCACHE_MEMORY`, this should still be fine until application decides to turn on the `NOCACHE` option. Right?

krish2718 replied:
Sure, just want to make sure there are no surprises to the current behavior.

jukkar commented on modules/nrf_wifi/os/shim.c:
```diff
@@ -20,28 +20,87 @@
 static void *zep_shim_mem_alloc(size_t size)
 {
-	size_t size_aligned = ROUND_UP(size, 4);
+	size = (size + 4) & 0xfffffffc;
```
Why not use the ROUND_UP macro here and below?

rado17 replied:
Done

krish2718 commented on drivers/wifi/nrf_wifi/Kconfig.nrfwifi:
```diff
@@ -535,6 +529,16 @@ config NRF70_RSSI_STALE_TIMEOUT_MS
 # Finetune defaults for certain system components used by the driver
+
+config NRF_WIFI_CTRL_HEAP_SIZE
+	int "Dedicated memory pool for control path"
```
```suggestion
	int "Dedicated memory pool for control plane"
```

rado17 replied:
Done.

rado17 committed:
modules: nrf_wifi: Include OSAL lib in buslib compilation

With introduction of OSAL wrapper for memory allocation, usage of
zephyr's native malloc calls need to be removed. When NRF70 driver
is not enabled, OSAL needs to be built explicitly.

Signed-off-by: Ravi Dondaputi <ravi.dondaputi@nordicsemi.no>
